src/main/scala/com/fulcrumgenomics/commons/util/ConfigurationLike.scala

```diff
@@ -0,0 +1,247 @@
+import scala.reflect.runtime.universe.{TypeTag, typeOf}
+import scala.util.{Failure, Success}
+
+/**
+  * Trait that provides useful methods for resolving all kinds of things in configuration.  Uses
+  * [[Config]] to retrieve configuration values in a type-safe way.
+  *
+  * Keeps track of all configuration keys that have been requested so that they can be reported later if desired.
+  *
+  * @example Retrieve a value at a given path, where the path must exist:
+  * {{{
+  *   scala> import com.fulcrumgenomics.commons.util.ConfigurationLike
+  *   scala> import com.typesafe.config.{Config, ConfigFactory}
+  *   scala> val conf = new ConfigurationLike {
+  *     val config: Config = ConfigFactory.parseString("""a.str = "string", a.int = 2, a.set = [1,2,3]""")
+  *   }
+  *   scala> conf[String]("a.str")
+  *   res0: String = string
+  *   scala> conf[Int]("a.int")
+  *   res1: Int = 2
+  *   scala> conf[Set[Int]]("a.set")
+  *   res2: Set[Int] = Set(1, 2, 3)
+  *   scala> conf[Set[Int]]("a.does-not-exist")
+  *   java.lang.IllegalArgumentException: Exception retrieving configuration key 'a.does-not-exist': No configuration setting found for key 'a.does-not-exist'
+  *   ...
+  * }}}
+  *
+  * @example Retrieve a value at a given path, where the path may not exist
+  * {{{
+  *   scala> import com.fulcrumgenomics.commons.util.ConfigurationLike
+  *   scala> import com.typesafe.config.{Config, ConfigFactory}
+  *   scala> val conf = new ConfigurationLike {
+  *     val config: Config = ConfigFactory.parseString("""a.str = "string", a.int = 2, a.set = [1,2,3]""")
+  *   }
+  *   scala> conf.get[String]("a.str")
+  *   res0: Option[String] = Some(string)
+  *   scala> conf.get[Int]("a.int")
+  *   res1: Option[Int] = Some(2)
+  *   scala> conf.get[Set[Int]]("a.set")
+  *   res2: Option[Set[Int]] = Some(Set(1, 2, 3))
+  *   scala> conf.get[Set[Int]]("a.does-not-exist")
+  *   res3: Option[Set[Int]] = None
+  * }}}
+  *
+  * @example Retrieve a value at a given path if it exists, otherwise return a default:
+  * {{{
+  *   scala> import com.fulcrumgenomics.commons.util.ConfigurationLike
+  *   scala> import com.typesafe.config.{Config, ConfigFactory}
+  *   scala> val conf = new ConfigurationLike {
+  *     val config: Config = ConfigFactory.parseString("""a.str = "string", a.int = 2, a.set = [1,2,3]""")
+  *   }
+  *   scala> conf.getOrElse[String]("a.does-not-exist", "default")
+  *   res0: String = default
+  *   scala> conf.getOrElse[Int]("a.does-not-exist", 0)
+  *   res1: Int = 0
+  *   scala> conf.getOrElse[Set[Int]]("a.does-not-exist", Set.empty)
+  *   res2: Set[Int] = Set()
+  * }}}
+  */
+trait ConfigurationLike {
+  /** The configuration to query. */
+  protected def config : Config
+
+  /** A logger to which to report exceptions. */
+  protected def _logger: Option[Logger] = None
+
+  /** The keys requested, in order. */
+  private val _requestedKeys = new java.util.concurrent.ConcurrentSkipListSet[String]()
```

Comment:
This used to be in dagr's `Configuration` singleton, so that it tracked the requested keys globally across all users of config regardless of where the traits were mixed in.  I don't think it's helpful to have it in the trait where it will only track the requests from each _instance_ that it's mixed into.  Maybe this can just go away here?